### PR TITLE
[16.0][MIG] account_invoice_triple_discount: Migration to 16.0

### PR DESCRIPTION
**Pull request by ramiadavid**
Migration to v16

#1250

### CODE REVIEW

**Comment by legalsylvain** [comment]
Hi. Thanks for porting this module. 
I have a functional question. In my company I receive many supplier invoices with Discount 1 & 2. (Odoo 12.)
Some computation are good (I mean Odoo computation = supplier computation), but some computation are bad. That is due to the fact that the supplier makes intermediate rounding.
Do you think, it could be possible to introduce some computation option (depending on the supplier ?)

If yes, I can try to implement that extra feature.

Note : I think that this module is very complex. We should try to talk to the Odoo SA dev team in the next odoo experience to try to have some hook to avoid all this code. Don't you think ?

Kind regard.

**Comment by ramiadavid** [comment]
> Hi. Thanks for porting this module. I have a functional question. In my company I receive many supplier invoices with Discount 1 & 2. (Odoo 12.) Some computation are good (I mean Odoo computation = supplier computation), but some computation are bad. That is due to the fact that the supplier makes intermediate rounding. Do you think, it could be possible to introduce some computation option (depending on the supplier ?)
> 
> If yes, I can try to implement that extra feature.
> 
> Note : I think that this module is very complex. We should try to talk to the Odoo SA dev team in the next odoo experience to try to have some hook to avoid all this code. Don't you think ?
> 
> Kind regard.

Hi @legalsylvain
In this module there is a bug in the calculation of discounts, it is discussed here #1240 , and I have uploaded a PR with the fix, I am waiting for it to be merged

Does this solve your problem?

**Comment by legalsylvain** [comment]
Hi @ramiadavid . thanks for your answer. it's différent topic. I talk about the fact that 2 results are valid depending on the supplier computation method.

**Review by FernandoRomera — APPROVED** [review]
Review, OK.

**Review by MRomeera — APPROVED** [review]
ok

**Review by AntoniRomera — APPROVED** [review]

**Review by sbejaoui — CHANGES_REQUESTED** [review]
Hi, 

Functional test

The invoice total don't match the lines subtotal.

image

**Comment by ramiadavid** [comment]
@sbejaoui @AntoniRomera @FernandoRomera @MRomeera 

This is fixed, but I see that the `discount_2` and `discount_3` fields are only shown if we add the 'Technical / Discount on lines' group, while the `discount` field is always available, likewise the discount field is set to optional="hide" while than the other two as optional="show" and in the `discount` field the name is abbreviated "Disc." while in the others it says "Discount"

Wouldn't it be better to unify these things to follow the same criteria as the original field?

**Comment by sbejaoui** [comment]
> Wouldn't it be better to unify these things to follow the same criteria as the original field?

+1, 
Thank you

**Comment by FernandoRomera** [comment]
@ramiadavid 
In account_invoice_triple_discount/report/invoice.xml, the names are:
                <span>Disc.2 %</span>
                <span>Disc.3 %</span>
without so many blanks and parentheses

>

**Comment by ramiadavid** [comment]
> @ramiadavid In account_invoice_triple_discount/report/invoice.xml, the names are: Disc.2 % Disc.3 % without so many blanks and parentheses
> 
> >

You're right, I've already changed it.

File: account_invoice_triple_discount/views/account_move.xml
Comment (lmignon) [inline account_invoice_triple_discount/views/account_move.xml:12]: @ramiadavid If you change the label, you need to update the translation files ..

**Review by lmignon — CHANGES_REQUESTED** [review]
LGTM (Code review + functional) but the source term into the translation files must be updated with the new labels for the discount columns....

File: account_invoice_triple_discount/views/account_move.xml
Comment (rousseldenis) [reply account_invoice_triple_discount/views/account_move.xml:12]: @lmignon This should be done once this is merged through weblate.

File: account_invoice_triple_discount/models/account_move_line.py
Comment (rousseldenis) [inline account_invoice_triple_discount/models/account_move_line.py]:
```suggestion
        records = super().create(values_list)
```

File: account_invoice_triple_discount/models/account_move_line.py
Comment (rousseldenis) [inline account_invoice_triple_discount/models/account_move_line.py]:
```suggestion
        res = super()._compute_totals()
```

File: account_invoice_triple_discount/models/account_move_line.py
Comment (rousseldenis) [inline account_invoice_triple_discount/models/account_move_line.py]:
```suggestion
        res = super()._compute_all_tax()
```

File: account_invoice_triple_discount/views/account_move.xml
Comment (lmignon) [reply account_invoice_triple_discount/views/account_move.xml:12]: @rousseldenis In this case, the source term is changed. That doesn't mean that we want to loose all the translations of the previous source term. That's why the translation files must be updated also replace the previous label by the new label for the source term.

Comment (ramiadavid) [reply account_invoice_triple_discount/views/account_move.xml:12]: @rousseldenis Updated translations for translated langs

**Review by lmignon — APPROVED** [review]
Thank you @ramiadavid for all this work. LGTM (Code review + functional)

**Review by rousseldenis — APPROVED** [review]

File: account_invoice_triple_discount/models/account_move_line.py
Comment (ramiadavid) [reply account_invoice_triple_discount/models/account_move_line.py]: done

File: account_invoice_triple_discount/models/account_move_line.py
Comment (ramiadavid) [reply account_invoice_triple_discount/models/account_move_line.py]: done

File: account_invoice_triple_discount/models/account_move_line.py
Comment (ramiadavid) [reply account_invoice_triple_discount/models/account_move_line.py]: done

**Comment by lmignon** [comment]
@OCA/accounting-maintainers Someone to merge this one?

**Comment by lmignon** [comment]
Thank you @adrienpeiffer for the merge!

**Pull request merged**
